Add pdf py proxy endpoints

hmstepanek opened the pull request:
Fix for https://github.com/hypothesis/proxy-server/issues/19. Depends on https://github.com/hypothesis/py-proxy/pull/1.

BigBlueHat commented:
For the record, :octocat:'s confused here...

The 8744655 that's mentioned on this PR was not the "ultimate" merging commit; https://github.com/hypothesis/via/commit/f2a411f7052de1c3e37036c090c262cf0939ac44 was.

https://github.com/hypothesis/via/commit/f2a411f7052de1c3e37036c090c262cf0939ac44 includes a revert of those debug statements--as we didn't want them in production.

Thanks for the upgrades and the PR, @ikreymer. The above is just a clerical note. :wink:

Commit:
Add pdf proxy endpoints

seanh commented on py_proxy/app.py:
```diff
@@ -9,13 +11,22 @@ def settings():
-    result = {}
+    embed_url = os.environ.get("H_EMBED_URL", "https://hypothes.is/embed.js")
+    nginx_server = os.environ.get("NGINX_SERVER", "https://via2.hypothes.is")
+    via_url = os.environ.get("VIA_URL", "https://via.hypothes.is")
```
Can we rename this one to `legacy_via_url` / `LEGACY_VIA_URL`?

seanh replied:
Not critical but I think we might want to remove the default values from these three settings and just let it crash on startup if any of the three envvars isn't set.

In practice these envvars always will be set in the dev, QA and prod environments so the defaults will never be used. If the app was ever deployed somewhere without these envvars set, I'm not sure that we'd want it to silently default to the production values, rather than crashing and telling the user they need to set the envvars.

Not critical though

hmstepanek replied:
I've added a check in there to raise a value error if a setting has no value. I thought this was more informative than letting the app crash in unexpected ways later on.

seanh replied:
:+1:

seanh commented on py_proxy/static/scripts/pdfjs-init.js:
```diff
@@ -0,0 +1,83 @@
+  appOptions.set('defaultUrl', '');
+
+  // Read configuration rendered into template as global vars.
+  var url = window.VIA_DOCUMENT_URL;
```
Can we rename this to something like:

* `PDF_URL`
* `PDF_DOWNLOAD_URL`
* `PROXIED_PDF_URL`
* `PROXIED_PDF_DOWNLOAD_URL`

It's the URL for PDF.js to download the PDF through py-proxy's static files proxy, so it doesn't actually have anything to do with Via.

We should avoid the word Via anyway since it's overloaded (unclear whether it refers to legacy Via or whether it's referring to the new proxy as Via) except when specifically referencing legacy Via and then we should explicitly say "legacy Via" (or `legacy_via` etc) to make it clear.

seanh commented on py_proxy/app.py:
```diff
@@ -9,13 +11,22 @@ def settings():
 
     """
-    result = {}
+    embed_url = os.environ.get("H_EMBED_URL", "https://hypothes.is/embed.js")
```
Can we rename this to `client_embed_url` and `CLIENT_EMBED_URL`? It doesn't actually have anything to do with h (the fact that it's actually h that serves the `/embed.js` URL is just an architectural oddity / hangover from when h and the client were not separate apps)

seanh commented on py_proxy/app.py:
```diff
@@ -9,13 +11,22 @@ def settings():
     """
-    result = {}
+    embed_url = os.environ.get("H_EMBED_URL", "https://hypothes.is/embed.js")
+    nginx_server = os.environ.get("NGINX_SERVER", "https://via2.hypothes.is")
```
So the intention is to deploy this app to `https://via2.hypothes.is`? I'm wondering how that will work since https://github.com/hypothesis/proxy-server is currently deployed to `https://via2.hypothes.is`?

hmstepanek replied:
Should be via3-just a mistake on my part.

seanh commented on py_proxy/static/scripts/pdfjs-init.js:
```diff
@@ -0,0 +1,83 @@
+
+  // Read configuration rendered into template as global vars.
+  var url = window.VIA_DOCUMENT_URL;
+  var clientEmbedUrl = window.VIA_H_EMBED_URL;
```
Rename to `window.CLIENT_EMBED_URL` to match the change I suggested to the corresponding envvar name above. `clientEmbedUrl` is fine for the JavaScript variable name.

seanh commented on py_proxy/templates/index.html.jinja2:
```diff
@@ -1,6 +1,99 @@
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <title>via.hypothes.is</title>
```
Probably not the best title since this isn't Via

seanh commented on py_proxy/templates/index.html.jinja2:
```diff
@@ -1,6 +1,99 @@
+           autofocus
+           class="url-field"
+           name="search"
+           placeholder="Paste a link to create or view annotations"
```
```suggestion
           placeholder="Paste a link to annotate"
```

Can we sneakily change this? I've always disliked this awkward phrasing in Via. I know you can technically just view annotations but I don't think it's necessary to say so here

hmstepanek replied:
Fine with me.

seanh commented on py_proxy/templates/index.html.jinja2:
```diff
@@ -1,6 +1,99 @@
+           placeholder="Paste a link to create or view annotations"
+           type="url"
+           >
+    <button type="submit" class="annotate-btn">Go!</button>
```
```suggestion
    <button type="submit" class="annotate-btn">Annotate!</button>
```

Another sneaky change to Via's UI. "Go" is not a good label for a button ("Go" could mean literally anything)

seanh replied:
This would also require a small CSS fix though. Not critical

seanh commented on py_proxy/views.py:
```diff
@@ -14,8 +16,54 @@ def status(_request):
+    """Serve favicon.ico file."""
+    return response.FileResponse(
+        "py_proxy/static/favicon.ico", request=request, content_type="image/x-icon"
+    )
```
Are these favicon and robots views necessary? I think the `config.add_static_view(name="static", path="static")` means they'll be served automatically

hmstepanek replied:
You are correct that they are served at /static/robots.txt however, they need to be served from the root endpoint (/robots.txt) so that line by itself will not work for either of these files.

seanh commented on py_proxy/views.py:
```diff
@@ -14,8 +16,54 @@ def status(_request):
+    renderer="py_proxy:templates/pdfjs_viewer.html.jinja2", route_name="pdf"
+)
+def pdf(request):
+    """Serve html page with client and the pdf embeded."""
```
```suggestion
    """An HTML page with client and the PDF embedded."""
```

seanh commented on tests/unit/py_proxy/views_test.py:
```diff
@@ -26,5 +29,115 @@ class TestIncludeMe:
+    #    result = views.pdf(pyramid_request)
+    #    import pdb; pdb.set_trace()
+
+    #    assert "pdfjs_viewer.html" in pyramid_request.override_renderer
```
We shouldn't commit any commented-out code

hmstepanek replied:
This is my placeholder for the jinja template rendering test. I was hoping to get it committed last night but it didn't happen-I have it locally now though. I'll commit it shortly.

seanh reviewed (changes requested):
A few minor suggestions, overall looks great! :+1:

It doesn't seem to work for me in testing though: PDF.js fails to load either the PDF or the client, and I see a JavaScript crash in the console

seanh commented:
Oh, looks like a `tests/.DS_Store` got committed

hmstepanek commented:
Maybe you can show me what you are seeing tomorrow with the crash? It's working for me locally. Note you must run ```make services``` first-maybe that's what you're missing? I do see a pdf.js error of some kind in the console despite the pdf loading which I may need Rob's input on as it's a bit hazy to me what's going on there.

I noticed there were some missing pdf.js files so perhaps those were the issue? Let me know if you still see an issue!

seanh commented:
Just re-tested with the latest commit and it's working for me now for proxying a simple PDF.

I don't think everything is working yet:

1. [This PDF](https://readthedocs.org/projects/h/downloads/pdf/latest/) doesn't seem to work for me. PDF.js displays an error. Possibly because that URL returns a redirect? But a simple non-redirecting PDF does seem to work

2. I haven't tested it in Canvas or with Canvas files or Google Drive files

I think we can merge this PR though, and can also go ahead with getting this deployed, and deal with remaining issues separately

hmstepanek commented:
Note I have adjusted the lint config for the tests to not lint for unused arguments and removed the pylint ignore comments from the tests where fixtures are passed in per our slack discussion [here](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1576217456016100).
